### ISSUE TITLE
JRuby install of 9.1.5.0 takes ages

### ISSUE THREAD

**Issue by perlun** [travis-ci/travis-ci#6892]
I wrote this as an issue a long time ago: https://github.com/travis-ci/travis-rubies/issues/22, and since then 9.1.6.0 has been released. It takes about 2-3 minutes to install it on Travis, far from acceptable.

Could we get a rvm-preinstalled 9.1.6.0 in the base image? Would be **extremely** appreciated.

**Comment by BanzaiMan** [comment]
If you are using GCE, please install HAVEGED.

**Comment by perlun** [comment]
It helps a bit, but not enough. I want it to be as fast as using MRI (i.e. a few seconds is acceptable).

**Comment by perlun** [comment]
(https://github.com/travis-ci/docs-travis-ci-com/pull/743 talks about haveged btw)

**Comment by BanzaiMan** [comment]
We will be updating pre-installed JRuby versions on GCE images, but there will always be a chance that your desired version is not (because we cannot pre-install all language runtimes).

**Comment by BanzaiMan** [comment]
I don't think "a few seconds" is achievable with JRuby installations.

**Comment by headius** [comment]
@BanzaiMan I came here to file a bug about this same problem and suggest that Travis add haveged to the base image. Is that a reasonable path forward? Container-based setups are always going to burn through entropy, so this seems like a good thing to do in general.

**Comment by headius** [comment]
FWIW, here's installing JRuby 9.1.6.0 [without haveged](https://travis-ci.org/ruby/psych/builds/176777492) and [with haveged](https://travis-ci.org/ruby/psych/builds/176780093). Seems like it's still a good idea to have it built-in.

**Comment by BanzaiMan** [comment]
@headius yes, I think it is in everyone's interest to have HAVEGED available by default.

**Comment by perlun** [comment]
> I don't think "a few seconds" is achievable with JRuby installations.

I'm very sure it is, at least when talking about "already installed" versions.

I did a quick test now, made a simple repo that uses a preinstalled version of JRuby: https://travis-ci.org/perlun/jruby-1.7-test/builds/177978599

The RVM install took 0.13s.

The problem there is that the only 9k JRuby available on the 12.04 image is `jruby-9.0.0.0.pre1`, i.e. _age-old_. I think the Trusty image have a somewhat more recent version, but we really need to get something from the JRuby 9.1.x series to be anywhere near usable.

(Sorry to be so complaining; we are paying every month both for the Travis service and now unfortunately a lot of waiting time on the builds, so that's why I'm quite annoyed with the current situation.)

**Comment by BanzaiMan** [comment]
I think you misunderstood my previous statement.

If a JRuby runtime is pre-installed, RVM is not installing anything. It is not installation taking "a few seconds". It's basically a no-op.

I understand the frustration, but, as I said, under the circumstances, I do not think "a few seconds" is achievable with JRuby installations. We will work on getting HAVEGED pre-installed to help.

**Comment by perlun** [comment]
> I understand the frustration, but, as I said, under the circumstances, I do not think "a few seconds" is achievable with JRuby installations. We will work on getting HAVEGED pre-installed to help.

I understand, and that's fine. But I also think we should add a non-beta version of jruby 9k (9.1.5.0 or 9.1.6.0) to either the 12.04 (current default) image or the new Trusty beta, as a preinstalled version. That would be the real "win" here, the haveged stuff is of course _good_ but it's still much much too slow IMHO. If there is another place to give feedback on the Trusty beta, let me know - I'm happy to reroute this as needed to get someone to dig into it.

Right now, there are many more MRI versions available (on 12.04) than JRuby versions which is problematic for us who work more with JRuby than MRI. This is the output from the linked build above:

```
$ rvm list

rvm rubies

   jruby-1.7.19-d18 [ x86_64 ]
=> jruby-1.7.19-d19 [ x86_64 ]
   jruby-9.0.0.0.pre1 [ x86_64 ]
   ree-1.8.7-2012.02 [ x86_64 ]
   ruby-1.8.7-head [ x86_64 ]
   ruby-1.8.7-p374 [ x86_64 ]
   ruby-1.9.2-p330 [ x86_64 ]
 * ruby-1.9.3-p551 [ x86_64 ]
   ruby-2.0.0-p598 [ x86_64 ]
   ruby-2.1.2 [ x86_64 ]
   ruby-2.1.3 [ x86_64 ]
   ruby-2.1.4 [ x86_64 ]
   ruby-2.1.5 [ x86_64 ]
   ruby-2.2.0 [ x86_64 ]
```

9.0.0.0.pre1 was released about _22 months_ ago (2015-01-20). It's practically unusable for most real-world scenarios, at least at my employer. Please fix. 😄

**Comment by nateberkopec** [comment]
Well, to be fair, the latest MRI version in that least was released in December 2014. Maybe @BanzaiMan can explain why *all* of the preinstalled ruby versions are old.

**Comment by perlun** [comment]
> Well, to be fair, the latest MRI version in that least was released in December 2014

True, but it's less of a problem with MRI since the install takes about 20s extra (compared to 2-3 minutes with JRuby). But I agree, it would be interesting to hear more details on why all preinstalled Rubies are ancient.

**Comment by BanzaiMan** [comment]
That is an excellent question!

I agree that these runtimes on Precise are ancient—practically dead in Internet time.

Many things contributed to delay in image updates. We moved our infrastructures to a new provider, we added Trusty support, and scaling things, etc. We focused our work on making updating images easier. That work is almost done.

We are planning image updates as we speak. It is not yet clear which ones we are going to have pre-installed (not many, unfortunately, due to disk space concerns), but a modern JRuby 9.1.x will be there, along with MRI versions.

**Comment by perlun** [comment]
> We are planning image updates as we speak. It is not yet clear which ones we are going to have pre-installed (not many, unfortunately, due to disk space concerns), but a modern JRuby 9.1.x will be there, along with MRI versions.

@BanzaiMan - thanks a lot for the status update on that. Eagerly waiting - keep us posted if you have something (even of alpha/beta quality) that you want use to test before it's rolled out to the general public.

**Comment by BanzaiMan** [comment]
You can look at https://github.com/travis-ci/packer-templates to get ideas of what's coming down the pipeline. The announcements go to https://docs.travis-ci.com/user/build-environment-updates/.

**Comment by perlun** [comment]
Any update on this one? It's equally frustrating still.

**Comment by perlun** [comment]
@BanzaiMan There hasn't been any update to the list of announcements for a full month. How are we doing with these changes? It's costing us a significant amount of time at the moment, in terms of meaningless waiting time for builds to complete...

**Comment by BanzaiMan** [comment]
@perlun Sorry for the delay. Precise images may not get the JRuby update love, given how old Precise itself is. https://github.com/travis-ci/packer-templates/issues/391 aims to get it on Trusty.

**Comment by perlun** [comment]
@BanzaiMan That's fine; we can move to Trusty once it's there. Thanks for the update! Awaiting this ❤️ from Travis to land. 😉

**Comment by BanzaiMan** [comment]
@perlun Thanks for your patience.

**Comment by perlun** [comment]
FWIW, Trusty seems to be much faster on JRuby installs in general: down to less than a minute (compared to 2-3 minutes normally) when enabling these settings:

```yml
sudo: false
dist: trusty
```

Still not "great" but at least a step in the right direction.

**Comment by perlun** [comment]
FWIW, I decided to implement this workaround for now: https://gist.github.com/perlun/a20924957203d4fc202e9b6523d79598

By skipping RVM and just downloading the binaries directly, we are down from 100-200 seconds to about 10s (including`gem install bundler`).

**Comment by perlun** [comment]
@BanzaiMan I looked at the list of preinstalled Ruby versions in one of our `trusty` builds. It seems like Ruby 2.2, 2.3 and 2.4 is preinstalled there, but _no JRuby version_. Can this really be the case? It would be super-nice if we could have jruby-latest and jruby-HEAD be preinstalled... because of the issues mentioned in this thread.
